File: backend/app/schemas/slide_storage.py
from pydantic import ConfigDict, BaseModel
from datetime import datetime
from typing import List, Optional

class UserRef(BaseModel):
    id: int
    full_name: str
    model_config = ConfigDict(from_attributes=True)

# --- Detail Schemas ---
class SlideStorageDetailBase(BaseModel):
    stain_id: int
    storage_location: Optional[str] = None
    remark: Optional[str] = None

class CaseShortInfo(BaseModel):
    accession_no: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True)

class SpecimenShortInfo(BaseModel):
    case: Optional[CaseShortInfo] = None
    
    model_config = ConfigDict(from_attributes=True)

class BlockShortInfo(BaseModel):
    block_code: Optional[str] = None
    specimen: Optional[SpecimenShortInfo] = None

    model_config = ConfigDict(from_attributes=True)

class TestShortInfo(BaseModel):
    id: int
    name: str
    category: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

class SlideShortInfo(BaseModel):
    id: int
    slide_no: int
    status: str
    block: Optional[BlockShortInfo] = None
    test: Optional[TestShortInfo] = None

    model_config = ConfigDict(from_attributes=True)

class CytoStainShortInfo(BaseModel):
    """Shared short info for GyneCytologyStain and NongyneCytologyStain."""
    id: int
    slide_no: int
    status: str
    case: Optional[CaseShortInfo] = None
    test: Optional[TestShortInfo] = None

    model_config = ConfigDict(from_attributes=True)

class SlideStorageDetailCreate(SlideStorageDetailBase):
    pass

class RunShortInfo(BaseModel):
    id: int
    run_no: str
    stain_category: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)

class SlideStorageDetail(SlideStorageDetailBase):
    id: int
    run_id: int
    stored_at: datetime
    stain: Optional[SlideShortInfo] = None
    gyne_stain: Optional[CytoStainShortInfo] = None
    nongyne_stain: Optional[CytoStainShortInfo] = None
    discard_status: bool = False
    discard_at: Optional[datetime] = None
    discard_by: Optional[UserRef] = None
    run: Optional[RunShortInfo] = None

    model_config = ConfigDict(from_attributes=True)

class SlideDisposeRequest(BaseModel):
    detail_ids: List[int]

# --- Run Schemas ---
class SlideStorageRunBase(BaseModel):
    run_no: Optional[str] = None
    user_id: int
    stain_category: Optional[str] = None
    remark: Optional[str] = None

class SlideStorageRunCreate(SlideStorageRunBase):
    pass

class SlideStorageRun(SlideStorageRunBase):
    id: int
    started_at: datetime
    finished_at: Optional[datetime] = None
    details: List[SlideStorageDetail] = []
    operator: Optional[UserRef] = None
    discard_status: bool = False
    discard_at: Optional[datetime] = None
    discard_by: Optional[UserRef] = None

    model_config = ConfigDict(from_attributes=True)

class SlideStorageDisposeRequest(BaseModel):
    run_ids: List[int]

class SlideStorageDetailUpdate(BaseModel):
    storage_location: Optional[str] = None
    remark: Optional[str] = None

class SlideStorageRunCreateBatch(BaseModel):
    user_id: int
    stain_category: Optional[str] = None
    items: List[SlideStorageDetailCreate]
    remark: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/slideStorage.ts
export type StainCategory = "HE" | "Special" | "IHC" | "Gyne" | "NonGyne";

export interface SlideStorageBatchPayload {
  user_id: number;
  stain_category?: StainCategory;
  items: { stain_id: number; storage_location?: string; remark?: string }[];
  remark?: string;
}

export interface SlideStorageRunResponse {
  id: number;
  run_no: string;
  started_at: string;
  finished_at?: string;
  user_id: number;
  operator?: { id: number; full_name: string };
  stain_category?: StainCategory;
  remark?: string;
  details: SlideStorageDetailResponse[];
  discard_status?: boolean;
  discard_at?: string;
  discard_by?: { id: number; full_name: string };
}

export interface SlideStorageDetailResponse {
  id: number;
  run_id: number;
  stain_id?: number;
  gyne_stain_id?: number;
  nongyne_stain_id?: number;
  storage_location?: string;
  remark?: string;
  stored_at: string;
  stain?: {
    slide_no?: string | number;
    test?: { name?: string };
    block?: { block_code?: string; specimen?: { case?: { accession_no?: string } } };
  };
  gyne_stain?: {
    slide_no?: string | number;
    test?: { name?: string };
    case?: { accession_no?: string };
  };
  nongyne_stain?: {
    slide_no?: string | number;
    test?: { name?: string };
    case?: { accession_no?: string };
  };
  run?: { id: number; run_no: string; stain_category?: string };
  discard_status?: boolean;
  discard_at?: string;
  discard_by?: { id: number; full_name: string };
}

export interface PendingStorageSlideNode {
  key: string | number;
  id: number;
  code: string;
  isCase: boolean;
  children?: PendingStorageSlideNode[];
}

export interface ScannedStorageSlide {
  id: number;
  code: string;
  accession_no: string;
  storage_location?: string;
  scannedAt: string;
}
